macros/helpers/prime_line.cfg

```diff
@@ -1,18 +1,46 @@
-    {% else %}
-      {% set prime_line_height = params.LINE_HEIGHT|default(0.6)|float %}
+    {% set prime_line_height = params.LINE_HEIGHT|default(printer["gcode_macro _USER_VARIABLES"].prime_line_height)|default(0.6)|float %}
+    {% set prime_line_adaptive = params.ADAPTIVE_MODE|default(True) %}
+    
+    # If the SIZE parameter is defined and not a dummy placeholder, we use it to do the adaptive bed mesh logic
+    {% set coordinatesFound = false %}
+    {% if params.SIZE is defined and params.SIZE != "0_0_0_0" %}
+        {% set xMinSpec, yMinSpec, xMaxSpec, yMaxSpec = params.SIZE.split('_')|map('trim')|map('int') %}
+        {% set coordinatesFound = true %}
+    {% elif printer.exclude_object is defined %}
+        {% if printer.exclude_object.objects %}
+            # Else if SIZE is not defined, we fallback to use the [exclude_object] tags
+            # This method is derived from Kyleisah KAMP repository: https://github.com/kyleisah/Klipper-Adaptive-Meshing-Purging)
+            {% set eo_points = printer.exclude_object.objects|map(attribute='polygon')|sum(start=[]) %}
+            {% set xMinSpec = eo_points|map(attribute=0)|min %}
+            {% set yMinSpec = eo_points|map(attribute=1)|min %}
+            {% set xMaxSpec = eo_points|map(attribute=0)|max %}
+            {% set yMaxSpec = eo_points|map(attribute=1)|max %}
+            {% set coordinatesFound = true %}
+        {% endif %}
     {% endif %}
 
-    # Set internal macro vars
+    # We get the default prime line position parameters
     {% set prime_line_x, prime_line_y = printer["gcode_macro _USER_VARIABLES"].prime_line_xy|map('float') %}
-    {% set prime_line_direction = printer["gcode_macro _USER_VARIABLES"].prime_line_direction|string|upper %}
+    {% set prime_line_x = params.START_X|default(prime_line_x)|float %}
+    {% set prime_line_y = params.START_Y|default(prime_line_y)|float %}
+    {% set prime_line_direction = params.LINE_DIRECTION|default(printer["gcode_macro _USER_VARIABLES"].prime_line_direction)|string|upper %}
+
+    # If first layer is retrieved and adaptive mode is enabled, then we replace the coordinates to do an adaptive purge
+    {% if coordinatesFound and prime_line_adaptive %}
+        {% set prime_line_margin = params.LINE_MARGIN|default(printer["gcode_macro _USER_VARIABLES"].prime_line_margin)|default(5.0)|float %}
+        {% set prime_line_x = [[prime_line_x, xMinSpec - prime_line_margin]|max, xMaxSpec + prime_line_margin]|min %}
+        {% set prime_line_y = [[prime_line_y, yMinSpec - prime_line_margin]|max, yMaxSpec + prime_line_margin]|min %}
+    {% endif %}
+
+    # Choose the way of printing the primeline (in + or -) alongside the direction to avoid going outside the bed boundaries
+    {% set axes_max = printer.toolhead.axis_maximum %}
+    {% set prime_line_way = -1 if (prime_line_direction == "X" and prime_line_x > axes_max.x/2) or (prime_line_direction == "Y" and prime_line_y > axes_max.y/2) else 1 %}
```

Comment:
Can we put the params.LINE_MARGIN line with other entry params for readability ?

As discuss with @elpopo-eng in https://github.com/Frix-x/klippain/pull/498#issuecomment-1963513548 there is an edge case where primeline could be outside the mesh. 
It add a bit of complixity but I suggest this code to mitigate this situation : 
```
# We find bed center : useful to determine pime_line position and prime_line_way
    {% set center_x, center_y  = [printer.toolhead.axis_maximum.x / 2, printer.toolhead.axis_maximum.y / 2]|map("float")  %}

    # If first layer is retrieved and adaptive mode is enabled, then we replace the coordinates to do an adaptive purge
    {% if if coordinatesFound and prime_line_adaptive" %}
        {% set min_x, min_y, max_x, max_y = fl_size.split("_")|map('float') %}

        # Fix prime_line start position while FL_SIZE is not in the same bed area
        {% set prime_line_x = 2*center_x - prime_line_x 
                               if (prime_line_x > center_x and xMaxSpec < center_x) or (prime_line_x < center_x and xMinSpec > center_x) 
                               else prime_line_x %}
        {% set prime_line_y = 2*center_y - prime_line_y 
                               if (prime_line_y > center_y and yMaxSpec < center_y) or (prime_line_y < center_y and yMinSpec > center_y) 
                               else prime_line_y %}

    
        {% set prime_line_x = [[prime_line_x, xMinSpec - prime_line_margin] | max, xMaxSpec + prime_line_margin] | min %}
        {% set prime_line_y = [[prime_line_y, yMinSpec - prime_line_margin] | max, yMaxSpec + prime_line_margin] | min %}
    {% endif %}
    
    # Choose the way of printing the primeline (in + or -) alongside the direction to avoid going outside the bed boundaries
    {% set prime_line_way = -1 if (prime_line_direction == "X" and prime_line_x > center_x) or (prime_line_direction == "Y" and prime_line_y > center_y) else 1 %}
```

Reply:
I fixed it using your code, thanks!